objects/src/accounts/delta/vault.rs

```diff
@@ -15,314 +18,466 @@ use crate::{
+    }
+
+    /// Returns an iterator over the (key, value) pairs of the map.
+    pub fn iter(&self) -> impl Iterator<Item = (&AccountId, &i64)> {
+        self.0.iter()
+    }
+
+    /// Merges another delta into this one, overwriting any existing values.
+    ///
+    /// The result is validated as part of the merge.
+    ///
+    /// # Errors
+    /// Returns an error if the result did not pass validation.
+    pub fn merge(&mut self, other: Self) -> Result<(), AccountDeltaError> {
+        // Merge fungible assets.
+        //
+        // Track fungible asset amounts - positive and negative. `i64` is not lossy while
+        // fungibles are restricted to 2^63-1. Overflow is still possible but we check for that.
+
+        for (&faucet_id, &amount) in other.0.iter() {
+            self.add_delta(faucet_id, amount)?;
         }
 
-        // make sure all added assets are unique
-        for (pos, asset) in self.added_assets.iter().enumerate() {
-            if self.added_assets[..pos].iter().any(|a| a.is_same(asset)) {
-                return Err(AccountDeltaError::DuplicateVaultUpdate(*asset));
-            }
+        self.validate()
+    }
+
+    // HELPER FUNCTIONS
+    // ---------------------------------------------------------------------------------------------
+
+    /// Updates the provided map with the provided key and amount. If the final amount is 0,
+    /// the entry is removed.
+    ///
+    /// # Errors
+    /// Returns an error if the delta would overflow.
+    fn add_delta(&mut self, faucet_id: AccountId, delta: i64) -> Result<(), AccountDeltaError> {
+        match self.0.entry(faucet_id) {
+            Entry::Vacant(entry) => {
+                entry.insert(delta);
+            },
+            Entry::Occupied(mut entry) => {
+                let old = *entry.get();
+                let new = old.checked_add(delta).ok_or(
+                    AccountDeltaError::FungibleAssetDeltaOverflow {
+                        faucet_id,
+                        this: old,
+                        other: delta,
+                    },
+                )?;
+
+                if new == 0 {
+                    entry.remove();
+                } else {
+                    *entry.get_mut() = new;
+                }
+            },
         }
 
-        // make sure all removed assets are the same
-        for (pos, asset) in self.removed_assets.iter().enumerate() {
-            if self.removed_assets[..pos].iter().any(|a| a.is_same(asset)) {
-                return Err(AccountDeltaError::DuplicateVaultUpdate(*asset));
-            }
+        Ok(())
+    }
 
-            if self.added_assets.iter().any(|a| a.is_same(asset)) {
-                return Err(AccountDeltaError::DuplicateVaultUpdate(*asset));
+    /// Checks whether this vault delta is valid.
+    ///
+    /// # Errors
+    /// Returns an error if one or more fungible assets' faucet IDs are invalid.
+    fn validate(&self) -> Result<(), AccountDeltaError> {
```

Comment:
Related to the previous comment: I think we need to run this logic only in the constructor. So, I would just move the code into the constructor and get rid of this function.

Reply:
I'd prefer to keep this function because it's self-explaining. Otherwise we would need to add comments in constructor that we do additional validation here. It's not a big deal and code would be a bit shorter, but IMO, moving code blocks to separate functions is a good practice.

Reply:
Sounds good. Let's keep it in a separate function.